theano/gpuarray/opt.py

```diff
@@ -158,7 +189,13 @@ def local_opt(node):
                 # the context was derived from the outputs
                 for i in node.inputs:
                     i.tag.context_name = context_name
-                new_op = maker(node, context_name)
+
+                try:
+                    new_op = maker(node.op, context_name, node.inputs)
+                except TypeError:
+                    # Pass the outputs so that the Local Optimizers don't need to
+                    # build the nodes again.
+                    new_op = maker(node.op, context_name, node.inputs, node.outputs)
```

Comment:
@abergeron told me he agree to make the outputs mandatory. So you can change all opt to take it and remove the try...except